### DISCUSSION TITLE
Do Google OR tools support sensitivity analysis?

### DISCUSSION THREAD

**Discussion by asmaier** [google/or-tools#3595] [Q&A] [answered]
After finding a solution I would like to do a sensitivity analysis on the results. E.g. I would like to get information about shadow prices and slack for the constraints. This is possible with python PuLp library as described at https://machinelearninggeek.com/sensitivity-analysis-in-python/ and also with gurobipy (see https://support.gurobi.com/hc/en-us/community/posts/360063478872-How-to-generate-sensitivity-analysis-report-in-detail). Can I do the same with Google OR Tools and if so, how? I couldn't find anything about that in the documentation.

**Comment by Regista6** [comment]
See [this](https://github.com/google/or-tools/blob/main/ortools/linear_solver/linear_solver.h#L1154) and [this](https://github.com/google/or-tools/blob/main/ortools/linear_solver/linear_solver.h#L1305). There is the option of getting dual_value as well as reduced_cost.

**Reply by Regista6** [reply]
I think dual_value calculates the shadow price.  For slack probably using [activities](https://github.com/google/or-tools/blob/2bb1fee017988cc84517dc6f368c958288989aba/ortools/linear_solver/linear_solver.h#L484) it can be calculated.

An [example](https://github.com/google/or-tools/blob/92eab40155d5566039c92662632f196174f8da47/examples/cpp/linear_programming.cc) that does all of the above.

**Reply by asmaier** [reply, marked as answer]
Thank you for your hints. It seems slack can be calculated by subtracting the activities from the upper bound of the constraint. With that I managed to convert the example from https://machinelearninggeek.com/sensitivity-analysis-in-python/ written in Python PuLP into Google OR Tools:

```py
import pandas as pd
from ortools.linear_solver import pywraplp
from ortools.init import pywrapinit

solver = pywraplp.Solver.CreateSolver('GLOP')
A = solver.NumVar(0.0, solver.infinity(), "A")
B = solver.NumVar(0.0, solver.infinity(), "B")
solver.Add(4 * A + 10 * B <= 100, "c0")
solver.Add(2 * A + 1 * B <= 22, "c1")
solver.Add(3 * A + 3 * B <= 39, "c2")

solver.Maximize(60 * A + 50 * B )

status = solver.Solve()
if status == pywraplp.Solver.OPTIMAL:
    print('Solution:')
    print('Objective value =', solver.Objective().Value())
    print('A =', A.solution_value())
    print('B =', B.solution_value())
else:
    print('The problem does not have an optimal solution.')
    
activities = solver.ComputeConstraintActivities()
o = [{'Name':c.name(), 'shadow price':c.dual_value(), 'slack': c.ub() - activities[i]} for i, c in enumerate(solver.constraints())]
print(pd.DataFrame(o))
```
Output is:
```
Solution:
Objective value = 740.0
A = 9.000000000000002
B = 3.9999999999999973
  Name  shadow price  slack
0   c0     -0.000000   24.0
1   c1     10.000000    0.0
2   c2     13.333333    0.0    
```
I think it would be nice if this or a similar example could be added to the Google OR Tools documentation. Not everybody is familiar with the fact that `shadow_price == dual_value` and `slack == upper_bound - activity`.

**Comment by asmaier** [comment]
Do I understand correctly: `constraint.dual_value()` will give me the shadow price? But what about the slack of a constraint?
